File: frontend/src/lib/simulator.ts
import { apiFetch } from "./api";

// ── Types ────────────────────────────────────────────────────────────────────

export interface VariablePattern {
  variable_key: string;
  pattern:
    | "sine"
    | "random_walk"
    | "step"
    | "ramp"
    | "counter"
    | "gps_track"
    | "noise"
    | "formula"
    | "csv_replay"
    | "manual";
  config: Record<string, unknown>;
}

export interface SimulatorConfig {
  id: number;
  name: string;
  description: string | null;
  template: string | null;
  device_id: number | null;
  device_uid: string | null;
  variable_patterns: VariablePattern[];
  interval_seconds: number;
  speed_multiplier: number;
  is_active: boolean;
  is_virtual_device: boolean;
  total_points_sent: number;
  started_at: string | null;
  last_value_at: string | null;
}

export interface SimulatorTemplate {
  id: string;
  name: string;
  description: string;
  icon: string;
  variable_patterns: VariablePattern[];
}

export interface SimulatorCreate {
  name: string;
  description?: string | null;
  template?: string | null;
  device_uid?: string | null;
  variable_patterns: VariablePattern[];
  interval_seconds?: number;
  speed_multiplier?: number;
  is_virtual_device?: boolean;
  auto_start?: boolean;
}

export interface SimulatorUpdate {
  name?: string;
  description?: string | null;
  variable_patterns?: VariablePattern[];
  interval_seconds?: number;
  speed_multiplier?: number;
}

// ── API ──────────────────────────────────────────────────────────────────────

const BASE = "/api/v1/simulator/configs";

export async function listSimulators(): Promise<SimulatorConfig[]> {
  return apiFetch<SimulatorConfig[]>(BASE);
}

export async function createSimulator(data: SimulatorCreate): Promise<SimulatorConfig> {
  return apiFetch<SimulatorConfig>(BASE, {
    method: "POST",
    body: JSON.stringify(data),
  });
}

export async function updateSimulator(id: number, data: SimulatorUpdate): Promise<SimulatorConfig> {
  return apiFetch<SimulatorConfig>(`${BASE}/${id}`, {
    method: "PUT",
    body: JSON.stringify(data),
  });
}

export async function deleteSimulator(id: number): Promise<void> {
  await apiFetch<void>(`${BASE}/${id}`, { method: "DELETE" });
}

export async function startSimulator(id: number): Promise<SimulatorConfig> {
  return apiFetch<SimulatorConfig>(`${BASE}/${id}/start`, { method: "POST" });
}

export async function stopSimulator(id: number): Promise<SimulatorConfig> {
  return apiFetch<SimulatorConfig>(`${BASE}/${id}/stop`, { method: "POST" });
}

export async function pulseValue(
  id: number,
  variableKey: string,
  value: number,
): Promise<void> {
  await apiFetch<void>(`${BASE}/${id}/pulse`, {
    method: "POST",
    body: JSON.stringify({ variable_key: variableKey, value }),
  });
}

export async function getTemplates(): Promise<SimulatorTemplate[]> {
  return apiFetch<SimulatorTemplate[]>("/api/v1/simulator/templates");
}

// ── Client-side pattern preview generators ──────────────────────────────────

/** Generate ~30s of sample data for client-side sparkline preview */
export function generatePreview(
  pattern: VariablePattern["pattern"],
  config: Record<string, unknown>,
  points = 60,
): { t: number; v: number }[] {
  const now = Date.now();
  const data: { t: number; v: number }[] = [];

  for (let i = 0; i < points; i++) {
    const t = now - (points - i) * 500;
    const sec = i * 0.5;
    let v = 0;

    switch (pattern) {
      case "sine": {
        const min = (config.min as number) ?? 0;
        const max = (config.max as number) ?? 100;
        const period = (config.period as number) ?? 60;
        const phase = (config.phase_offset as number) ?? 0;
        const mid = (min + max) / 2;
        const amp = (max - min) / 2;
        v = mid + amp * Math.sin((2 * Math.PI * sec) / period + phase);
        break;
      }
      case "random_walk": {
        const center = (config.center as number) ?? 50;
        const volatility = (config.volatility as number) ?? 5;
        const minB = (config.min_bound as number) ?? 0;
        const maxB = (config.max_bound as number) ?? 100;
        if (i === 0) v = center;
        else v = data[i - 1].v + (Math.random() - 0.5) * volatility;
        v = Math.max(minB, Math.min(maxB, v));
        break;
      }
      case "step": {
        const valuesStr = (config.values as string) ?? "0,50,100";
        const values = valuesStr
          .split(",")
          .map((s: string) => parseFloat(s.trim()))
          .filter((n: number) => !isNaN(n));
        const interval = (config.interval as number) ?? 5;
        if (values.length > 0) {
          const idx = Math.floor(sec / interval) % values.length;
          v = values[idx];
        }
        break;
      }
      case "ramp": {
        const start = (config.start as number) ?? 0;
        const end = (config.end as number) ?? 100;
        const duration = (config.duration as number) ?? 30;
        const loop = (config.loop as boolean) ?? true;
        const progress = loop ? (sec % duration) / duration : Math.min(sec / duration, 1);
        v = start + (end - start) * progress;
        break;
      }
      case "counter": {
        const startVal = (config.start as number) ?? 0;
        const increment = (config.increment as number) ?? 1;
        const resetAt = (config.reset_at as number) ?? 1000;
        v = (startVal + i * increment) % resetAt;
        break;
      }
      case "noise": {
        const center = (config.center as number) ?? 50;
        const amplitude = (config.amplitude as number) ?? 10;
        v = center + (Math.random() - 0.5) * 2 * amplitude;
        break;
      }
      case "formula": {
        try {
          const expr = (config.expression as string) ?? "sin(t)";
          const fn = new Function(
            "t",
            "sin",
            "cos",
            "random",
            "pi",
            "abs",
            "sqrt",
            `return (${expr})`,
          );
          v = fn(sec, Math.sin, Math.cos, Math.random, Math.PI, Math.abs, Math.sqrt) as number;
          if (!isFinite(v)) v = 0;
        } catch {
          v = 0;
        }
        break;
      }
      case "gps_track":
        // GPS generates lat/lng, just show a sine-like path for preview
        v = 48.2 + 0.01 * Math.sin((2 * Math.PI * sec) / 30);
        break;
      case "csv_replay":
        // CSV replay can't be previewed client-side; show placeholder
        v = 50 + 20 * Math.sin((2 * Math.PI * sec) / 20);
        break;
      case "manual":
        v = (config.initial_value as number) ?? 0;
        break;
      default:
        v = 0;
    }

    data.push({ t, v });
  }
  return data;
}

// ── Pattern metadata ────────────────────────────────────────────────────────

export const PATTERN_OPTIONS: {
  value: VariablePattern["pattern"];
  label: string;
  labelDe: string;
}[] = [
  { value: "sine", label: "Sine Wave", labelDe: "Sinuswelle" },
  { value: "random_walk", label: "Random Walk", labelDe: "Random Walk" },
  { value: "step", label: "Step", labelDe: "Stufe" },
  { value: "ramp", label: "Ramp", labelDe: "Rampe" },
  { value: "counter", label: "Counter", labelDe: "Zaehler" },
  { value: "gps_track", label: "GPS Track", labelDe: "GPS-Track" },
  { value: "noise", label: "Noise", labelDe: "Rauschen" },
  { value: "formula", label: "Custom Formula", labelDe: "Formel" },
  { value: "csv_replay", label: "CSV Replay", labelDe: "CSV-Wiedergabe" },
  { value: "manual", label: "Manual", labelDe: "Manuell" },
];

// ── Built-in templates (client-side fallback) ───────────────────────────────

export const BUILTIN_TEMPLATES: SimulatorTemplate[] = [
  {
    id: "temperature",
    name: "Temperature Sensor",
    description: "Simulates a room temperature sensor with natural fluctuations",
    icon: "thermometer",
    variable_patterns: [
      { variable_key: "temperature", pattern: "sine", config: { min: 18, max: 28, period: 300 } },
      { variable_key: "humidity", pattern: "random_walk", config: { center: 55, volatility: 2, min_bound: 30, max_bound: 80 } },
    ],
  },
  {
    id: "energy",
    name: "Energy Meter",
    description: "Simulates power consumption with daily cycles",
    icon: "bolt",
    variable_patterns: [
      { variable_key: "power_watts", pattern: "sine", config: { min: 200, max: 3500, period: 600 } },
      { variable_key: "energy_kwh", pattern: "counter", config: { start: 0, increment: 0.5, reset_at: 10000 } },
    ],
  },
  {
    id: "gps",
    name: "GPS Tracker",
    description: "Simulates a moving vehicle with GPS coordinates",
    icon: "map-pin",
    variable_patterns: [
      { variable_key: "latitude", pattern: "gps_track", config: { waypoints: "48.2082,16.3738;48.2100,16.3800;48.2050,16.3850" } },
      { variable_key: "longitude", pattern: "gps_track", config: { waypoints: "16.3738,48.2082;16.3800,48.2100;16.3850,48.2050" } },
      { variable_key: "speed_kmh", pattern: "random_walk", config: { center: 40, volatility: 8, min_bound: 0, max_bound: 120 } },
    ],
  },
  {
    id: "motion",
    name: "Motion Sensor",
    description: "Simulates a PIR motion detector with binary events",
    icon: "activity",
    variable_patterns: [
      { variable_key: "motion_detected", pattern: "step", config: { values: "0,0,0,1,0,0,1", interval: 10 } },
      { variable_key: "lux", pattern: "random_walk", config: { center: 300, volatility: 20, min_bound: 0, max_bound: 1000 } },
    ],
  },
  {
    id: "weather",
    name: "Weather Station",
    description: "Simulates outdoor weather with multiple sensors",
    icon: "cloud-sun",
    variable_patterns: [
      { variable_key: "temperature", pattern: "sine", config: { min: -5, max: 35, period: 600 } },
      { variable_key: "humidity", pattern: "random_walk", config: { center: 60, volatility: 3, min_bound: 20, max_bound: 95 } },
      { variable_key: "pressure_hpa", pattern: "sine", config: { min: 990, max: 1030, period: 900 } },
      { variable_key: "wind_speed", pattern: "noise", config: { center: 12, amplitude: 8 } },
    ],
  },
  // Sprint 8 R4 Bucket C F19: new templates that line up with the
  // expanded Sprint 5b hardware component catalog (BME280 / BH1750 /
  // HC-SR04 / servo / LED PWM / door contact). These round out the
  // hardware-demo story in the Sandbox.
  {
    id: "bme280",
    name: "BME280 Environmental",
    description: "BME280 I2C sensor — temperature, humidity, pressure, and IAQ gas resistance",
    icon: "thermometer",
    variable_patterns: [
      { variable_key: "temperature", pattern: "sine", config: { min: 19, max: 24.5, period: 300 } },
      { variable_key: "humidity", pattern: "sine", config: { min: 42, max: 58, period: 300, phase_offset: 3.14159 } },
      { variable_key: "pressure", pattern: "noise", config: { center: 1013.25, amplitude: 3.5 } },
      { variable_key: "gas_resistance_kohm", pattern: "random_walk", config: { center: 150, volatility: 12, min_bound: 50, max_bound: 400 } },
    ],
  },
  {
    id: "bh1750",
    name: "BH1750 Light Sensor",
    description: "BH1750 ambient light meter with a day/night lux cycle",
    icon: "sun",
    variable_patterns: [
      { variable_key: "lux", pattern: "sine", config: { min: 0, max: 1200, period: 300 } },
    ],
  },
  {
    id: "ultrasonic",
    name: "HC-SR04 Distance",
    description: "Ultrasonic distance sensor with presence spikes on top of a random-walk baseline",
    icon: "ruler",
    variable_patterns: [
      { variable_key: "distance_cm", pattern: "random_walk", config: { center: 85, volatility: 18, min_bound: 4, max_bound: 200 } },
      { variable_key: "object_detected", pattern: "step", config: { values: "0,0,1,0,0,0,1,0", interval: 15 } },
    ],
  },
  {
    id: "servo",
    name: "Servo Motor",
    description: "Servo actuator sweeping between set-points with a matching target channel",
    icon: "rotate-cw",
    variable_patterns: [
      { variable_key: "angle_deg", pattern: "sine", config: { min: 0, max: 180, period: 120 } },
      { variable_key: "target_angle_deg", pattern: "step", config: { values: "0,45,90,135,180,135,90,45", interval: 15 } },
    ],
  },
  {
    id: "led_pwm",
    name: "LED Dimmer (PWM)",
    description: "PWM LED channel — brightness ramps and duty-cycle pulses",
    icon: "lightbulb",
    variable_patterns: [
      { variable_key: "brightness_pct", pattern: "sine", config: { min: 0, max: 100, period: 300 } },
      { variable_key: "duty_cycle", pattern: "ramp", config: { start: 0, end: 255, duration: 600, loop: true } },
    ],
  },
  {
    id: "door_contact",
    name: "Door/Window Contact",
    description: "Reed-switch contact sensor — open/closed events plus battery drain",
    icon: "door-open",
    variable_patterns: [
      { variable_key: "contact_closed", pattern: "step", config: { values: "1,1,1,0,1,1,0,1", interval: 45 } },
      { variable_key: "battery_pct", pattern: "ramp", config: { start: 100, end: 0, duration: 2592000, loop: true } },
    ],
  },
  {
    id: "custom",
    name: "Custom",
    description: "Start from scratch with an empty simulator",
    icon: "wrench",
    variable_patterns: [],
  },
];

File: app/api/v1/simulator.py
"""Simulator API — create, manage, start/stop device simulators."""

from __future__ import annotations

import secrets
from datetime import datetime, timezone
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import select, delete
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_db
from app.api.deps_auth import get_current_user
from app.api.deps_org import get_current_org_id
from app.db.models.device import Device
from app.db.models.simulator import SimulatorConfig
from app.db.models.user import User
from app.db.models.variables import VariableDefinition
from app.schemas.simulator import (
    SimulatorCreate,
    SimulatorUpdate,
    SimulatorOut,
    SimulatorQuickPulse,
    TemplateInfo,
    VariablePattern,
)
from app.core import simulator_worker

router = APIRouter(prefix="/simulator", tags=["simulator"])


# ---------------------------------------------------------------------------
# Templates
# ---------------------------------------------------------------------------

_TEMPLATES: dict[str, TemplateInfo] = {
    "temperature": TemplateInfo(
        key="temperature",
        name="Temperature Sensor",
        description="Temperature, humidity, and pressure sensor with daily cycles",
        variable_patterns=[
            VariablePattern(
                variable_key="temperature",
                pattern="sine",
                config={"min": 18.0, "max": 28.0, "period_seconds": 86400},
            ),
            VariablePattern(
                variable_key="humidity",
                pattern="sine",
                config={"min": 40.0, "max": 70.0, "period_seconds": 86400, "phase_offset": 3.14159},
            ),
            VariablePattern(
                variable_key="pressure",
                pattern="noise",
                config={"center": 1013.0, "amplitude": 5.0},
            ),
        ],
    ),
    "energy": TemplateInfo(
        key="energy",
        name="Energy Meter",
        description="Power consumption, cumulative energy, and voltage readings",
        variable_patterns=[
            VariablePattern(
                variable_key="power_watts",
                pattern="sine",
                config={"min": 200.0, "max": 2000.0, "period_seconds": 86400},
            ),
            VariablePattern(
                variable_key="energy_kwh",
                pattern="counter",
                config={"start": 0.0, "increment": 0.5, "interval_seconds": 60},
            ),
            VariablePattern(
                variable_key="voltage",
                pattern="noise",
                config={"center": 230.0, "amplitude": 2.0},
            ),
        ],
    ),
    "gps": TemplateInfo(
        key="gps",
        name="GPS Tracker",
        description="Moving GPS position with speed and battery drain",
        variable_patterns=[
            VariablePattern(
                variable_key="position",
                pattern="gps_track",
                config={
                    "waypoints": [
                        {"lat": 52.5200, "lng": 13.4050},
                        {"lat": 52.5230, "lng": 13.4100},
                        {"lat": 52.5260, "lng": 13.4000},
                        {"lat": 52.5220, "lng": 13.3950},
                        {"lat": 52.5200, "lng": 13.4050},
                    ],
                    "speed_kmh": 30.0,
                },
            ),
            VariablePattern(
                variable_key="speed",
                pattern="noise",
                config={"center": 30.0, "amplitude": 10.0},
            ),
            VariablePattern(
                variable_key="battery",
                pattern="ramp",
                config={"start": 100.0, "end": 0.0, "duration_seconds": 28800, "loop": True},
            ),
        ],
    ),
    "motion": TemplateInfo(
        key="motion",
        name="Motion Sensor",
        description="Motion detection with ambient light levels",
        variable_patterns=[
            VariablePattern(
                variable_key="motion",
                pattern="step",
                config={"values": [False, False, False, True, False, True], "interval_seconds": 30},
            ),
            VariablePattern(
                variable_key="luminance",
                pattern="sine",
                config={"min": 0.0, "max": 1000.0, "period_seconds": 86400},
            ),
        ],
    ),
    "weather": TemplateInfo(
        key="weather",
        name="Weather Station",
        description="Full weather station with temperature, humidity, wind, and rain",
        variable_patterns=[
            VariablePattern(
                variable_key="temperature",
                pattern="sine",
                config={"min": 5.0, "max": 25.0, "period_seconds": 86400},
            ),
            VariablePattern(
                variable_key="humidity",
                pattern="sine",
                config={"min": 30.0, "max": 90.0, "period_seconds": 86400, "phase_offset": 3.14159},
            ),
            VariablePattern(
                variable_key="wind_speed",
                pattern="random_walk",
                config={"center": 10.0, "volatility": 2.0, "min_bound": 0.0, "max_bound": 50.0},
            ),
            VariablePattern(
                variable_key="rain_mm",
                pattern="counter",
                config={"start": 0.0, "increment": 0.2, "interval_seconds": 300, "reset_at": 50.0},
            ),
        ],
    ),
    # Sprint 8 R4 Bucket C F19: new simulator templates that line up with
    # the expanded Sprint 5b component catalog (BME280, BH1750, HC-SR04,
    # servo, LED PWM, door contact). These round out the "hardware demo"
    # story so the Sandbox looks alive after a fresh register.
    "bme280": TemplateInfo(
        key="bme280",
        name="BME280 Environmental",
        description="BME280 I2C sensor — temperature, humidity, pressure, and gas resistance (IAQ)",
        variable_patterns=[
            VariablePattern(
                variable_key="temperature",
                pattern="sine",
                config={"min": 19.0, "max": 24.5, "period_seconds": 86400},
            ),
            VariablePattern(
                variable_key="humidity",
                pattern="sine",
                config={"min": 42.0, "max": 58.0, "period_seconds": 86400, "phase_offset": 3.14159},
            ),
            VariablePattern(
                variable_key="pressure",
                pattern="noise",
                config={"center": 1013.25, "amplitude": 3.5},
            ),
            VariablePattern(
                variable_key="gas_resistance_kohm",
                pattern="random_walk",
                config={"center": 150.0, "volatility": 12.0, "min_bound": 50.0, "max_bound": 400.0},
            ),
        ],
    ),
    "bh1750": TemplateInfo(
        key="bh1750",
        name="BH1750 Light Sensor",
        description="BH1750 ambient light meter — day/night lux cycle with curtain noise",
        variable_patterns=[
            VariablePattern(
                variable_key="lux",
                pattern="sine",
                config={"min": 0.0, "max": 1200.0, "period_seconds": 86400},
            ),
        ],
    ),
    "ultrasonic": TemplateInfo(
        key="ultrasonic",
        name="HC-SR04 Distance",
        description="HC-SR04 ultrasonic distance sensor — ping-pong distance with presence spikes",
        variable_patterns=[
            VariablePattern(
                variable_key="distance_cm",
                pattern="random_walk",
                config={"center": 85.0, "volatility": 18.0, "min_bound": 4.0, "max_bound": 200.0},
            ),
            VariablePattern(
                variable_key="object_detected",
                pattern="step",
                config={"values": [False, False, True, False, False, False, True, False], "interval_seconds": 15},
            ),
        ],
    ),
    "servo": TemplateInfo(
        key="servo",
        name="Servo Motor",
        description="Servo actuator — angle sweeps between two set-points on schedule",
        variable_patterns=[
            VariablePattern(
                variable_key="angle_deg",
                pattern="sine",
                config={"min": 0.0, "max": 180.0, "period_seconds": 120},
            ),
            VariablePattern(
                variable_key="target_angle_deg",
                pattern="step",
                config={"values": [0, 45, 90, 135, 180, 135, 90, 45], "interval_seconds": 15},
            ),
        ],
    ),
    "led_pwm": TemplateInfo(
        key="led_pwm",
        name="LED Dimmer (PWM)",
        description="PWM LED channel — brightness ramps up and fades through the day",
        variable_patterns=[
            VariablePattern(
                variable_key="brightness_pct",
                pattern="sine",
                config={"min": 0.0, "max": 100.0, "period_seconds": 300},
            ),
            VariablePattern(
                variable_key="duty_cycle",
                pattern="ramp",
                config={"start": 0.0, "end": 255.0, "duration_seconds": 600, "loop": True},
            ),
        ],
    ),
    "door_contact": TemplateInfo(
        key="door_contact",
        name="Door/Window Contact",
        description="Reed-switch contact sensor — open/closed events with battery drain",
        variable_patterns=[
            VariablePattern(
                variable_key="contact_closed",
                pattern="step",
                config={"values": [True, True, True, False, True, True, False, True], "interval_seconds": 45},
            ),
            VariablePattern(
                variable_key="battery_pct",
                pattern="ramp",
                config={"start": 100.0, "end": 0.0, "duration_seconds": 86400 * 30, "loop": True},
            ),
        ],
    ),
    "custom": TemplateInfo(
        key="custom",
        name="Custom",
        description="Empty template — define your own variables and patterns",
        variable_patterns=[],
    ),
}


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _sim_to_out(sim: SimulatorConfig) -> SimulatorOut:
    patterns = sim.variable_patterns or []
    parsed_patterns = []
    for p in patterns:
        try:
            parsed_patterns.append(VariablePattern(**p))
        except Exception:
            parsed_patterns.append(
                VariablePattern(variable_key=p.get("variable_key", "?"), pattern="noise", config={})
            )

    return SimulatorOut(
        id=sim.id,
        org_id=sim.org_id,
        owner_id=sim.owner_id,
        device_id=sim.device_id,
        device_uid=sim.device_uid,
        name=sim.name,
        description=sim.description,
        template=sim.template,
        variable_patterns=parsed_patterns,
        interval_seconds=sim.interval_seconds,
        speed_multiplier=sim.speed_multiplier,
        is_active=sim.is_active,
        is_virtual_device=sim.is_virtual_device,
        total_points_sent=sim.total_points_sent,
        started_at=sim.started_at,
        last_value_at=sim.last_value_at,
        created_at=sim.created_at,
        updated_at=sim.updated_at,
    )


async def _get_sim_or_404(
    db: AsyncSession, sim_id: int, owner_id: int
) -> SimulatorConfig:
    res = await db.execute(
        select(SimulatorConfig).where(
            SimulatorConfig.id == sim_id,
            SimulatorConfig.owner_id == owner_id,
        )
    )
    sim = res.scalar_one_or_none()
    if sim is None:
        raise HTTPException(status_code=404, detail="simulator not found")
    return sim


async def _create_virtual_device(
    db: AsyncSession,
    user: User,
    org_id: int | None,
    sim_name: str,
    variable_patterns: list[VariablePattern],
) -> Device:
    """Create a virtual device and its variable definitions."""
    uid = f"sim-{sim_name.lower().replace(' ', '-')[:30]}-{secrets.token_hex(4)}"
    device = Device(
        device_uid=uid,
        name=f"[SIM] {sim_name}",
        device_type="simulator",
        category="hardware",
        owner_user_id=user.id,
        org_id=org_id,
        is_claimed=True,
        capabilities={"simulated": True},
        last_seen_at=datetime.now(timezone.utc),
    )
    db.add(device)
    await db.flush()

    # Create variable definitions for the patterns
    for pat in variable_patterns:
        # Check if definition already exists
        existing = await db.execute(
            select(VariableDefinition).where(VariableDefinition.key == pat.variable_key)
        )
        if existing.scalar_one_or_none() is None:
            vtype = "float"
            if pat.pattern == "gps_track":
                vtype = "json"
            elif pat.pattern == "step":
                vals = pat.config.get("values", [])
                if vals and isinstance(vals[0], bool):
                    vtype = "bool"
            db.add(VariableDefinition(
                key=pat.variable_key,
                scope="device",
                value_type=vtype,
                description=f"Simulated: {pat.variable_key}",
                device_writable=True,
            ))

    await db.flush()
    return device


# ---------------------------------------------------------------------------
# Templates endpoint
# ---------------------------------------------------------------------------


@router.get("/templates", response_model=list[TemplateInfo])
async def list_templates(
    _user: User = Depends(get_current_user),
):
    """List available simulator templates with predefined variable patterns."""
    return list(_TEMPLATES.values())


# ---------------------------------------------------------------------------
# CRUD
# ---------------------------------------------------------------------------


@router.get("/configs", response_model=list[SimulatorOut])
async def list_simulators(
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    """List all simulators owned by the current user."""
    res = await db.execute(
        select(SimulatorConfig)
        .where(SimulatorConfig.owner_id == user.id)
        .order_by(SimulatorConfig.id)
    )
    sims = res.scalars().all()
    return [_sim_to_out(s) for s in sims]


@router.post("/configs", response_model=SimulatorOut, status_code=201)
async def create_simulator(
    body: SimulatorCreate,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
    org_id: int | None = Depends(get_current_org_id),
):
    """Create a new simulator configuration.

    If is_virtual_device is true, a virtual device is created automatically.
    If device_uid is provided and is_virtual_device is false, the simulator attaches
    to the existing device.
    If a template is specified, its variable_patterns are used as defaults (unless
    the caller provides their own).
    """
    # Apply template patterns if none provided
    patterns = body.variable_patterns
    if not patterns and body.template and body.template in _TEMPLATES:
        patterns = _TEMPLATES[body.template].variable_patterns

    device_id: int | None = None
    device_uid: str | None = body.device_uid

    if body.is_virtual_device:
        device = await _create_virtual_device(db, user, org_id, body.name, patterns)
        device_id = device.id
        device_uid = device.device_uid
    elif body.device_uid:
        # Attach to existing device
        res = await db.execute(
            select(Device).where(
                Device.device_uid == body.device_uid,
                Device.owner_user_id == user.id,
            )
        )
        existing = res.scalar_one_or_none()
        if existing is None:
            raise HTTPException(status_code=404, detail="device not found")
        device_id = existing.id
        device_uid = existing.device_uid

    sim = SimulatorConfig(
        org_id=org_id,
        owner_id=user.id,
        device_id=device_id,
        device_uid=device_uid,
        name=body.name,
        description=body.description,
        template=body.template,
        variable_patterns=[p.model_dump() for p in patterns],
        interval_seconds=body.interval_seconds,
        speed_multiplier=body.speed_multiplier,
        is_virtual_device=body.is_virtual_device,
    )
    db.add(sim)
    await db.commit()
    await db.refresh(sim)
    return _sim_to_out(sim)


@router.get("/configs/{sim_id}", response_model=SimulatorOut)
async def get_simulator(
    sim_id: int,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    """Get a simulator configuration with current stats."""
    sim = await _get_sim_or_404(db, sim_id, user.id)
    # Sync is_active with worker state
    running = simulator_worker.is_running(sim.id)
    if sim.is_active != running:
        sim.is_active = running
        await db.commit()
        await db.refresh(sim)
    return _sim_to_out(sim)


@router.put("/configs/{sim_id}", response_model=SimulatorOut)
async def update_simulator(
    sim_id: int,
    body: SimulatorUpdate,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    """Update a simulator configuration. Stops the simulator if it's running."""
    sim = await _get_sim_or_404(db, sim_id, user.id)

    # Stop if running (config changed)
    if sim.is_active:
        simulator_worker.stop_simulator(sim.id)
        sim.is_active = False

    if body.name is not None:
        sim.name = body.name
    if body.description is not None:
        sim.description = body.description
    if body.variable_patterns is not None:
        sim.variable_patterns = [p.model_dump() for p in body.variable_patterns]
    if body.interval_seconds is not None:
        sim.interval_seconds = body.interval_seconds
    if body.speed_multiplier is not None:
        sim.speed_multiplier = body.speed_multiplier

    await db.commit()
    await db.refresh(sim)
    return _sim_to_out(sim)


@router.delete("/configs/{sim_id}", status_code=204)
async def delete_simulator(
    sim_id: int,
    delete_device: bool = Query(False, description="Also delete the virtual device"),
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    """Delete a simulator. Optionally deletes the associated virtual device."""
    sim = await _get_sim_or_404(db, sim_id, user.id)

    # Stop if running
    if sim.is_active:
        simulator_worker.stop_simulator(sim.id)

    virtual_device_id = sim.device_id if sim.is_virtual_device else None

    await db.execute(
        delete(SimulatorConfig).where(SimulatorConfig.id == sim.id)
    )

    # Optionally delete the virtual device
    if delete_device and virtual_device_id:
        await db.execute(
            delete(Device).where(Device.id == virtual_device_id)
        )

    await db.commit()


# ---------------------------------------------------------------------------
# Start / Stop / Pulse
# ---------------------------------------------------------------------------


@router.post("/configs/{sim_id}/start", response_model=SimulatorOut)
async def start_simulator(
    sim_id: int,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    """Start a simulator. Begins generating data in the background."""
    sim = await _get_sim_or_404(db, sim_id, user.id)

    if not sim.device_id or not sim.device_uid:
        raise HTTPException(status_code=400, detail="simulator has no device linked")

    if sim.is_active and simulator_worker.is_running(sim.id):
        return _sim_to_out(sim)  # Already running

    sim.is_active = True
    sim.started_at = datetime.now(timezone.utc)
    await db.commit()
    await db.refresh(sim)

    simulator_worker.start_simulator(sim.id)
    return _sim_to_out(sim)


@router.post("/configs/{sim_id}/stop", response_model=SimulatorOut)
async def stop_simulator(
    sim_id: int,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    """Stop a running simulator."""
    sim = await _get_sim_or_404(db, sim_id, user.id)

    simulator_worker.stop_simulator(sim.id)
    sim.is_active = False
    await db.commit()
    await db.refresh(sim)
    return _sim_to_out(sim)


@router.post("/configs/{sim_id}/pulse", response_model=dict)
async def pulse_single_value(
    sim_id: int,
    body: SimulatorQuickPulse,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    """Send a single value for one variable on the simulator's device."""
    sim = await _get_sim_or_404(db, sim_id, user.id)

    if not sim.device_id or not sim.device_uid:
        raise HTTPException(status_code=400, detail="simulator has no device linked")

    from app.api.v1.telemetry import _bridge_telemetry_to_variables

    await _bridge_telemetry_to_variables(
        device_id=sim.device_id,
        device_uid=sim.device_uid,
        event_type="simulator",
        payload={body.variable_key: body.value},
    )

    return {"ok": True, "variable_key": body.variable_key, "value": body.value}


# ---------------------------------------------------------------------------
# Quick Pulse (device-centric, no simulator required)
# ---------------------------------------------------------------------------


@router.post("/quick-pulse", response_model=dict)
async def quick_pulse(
    body: SimulatorQuickPulse,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    """Push a single value for any variable on any device owned by the user.

    Useful from Device Detail page to quickly inject test data without
    creating a full simulator configuration.
    """
    res = await db.execute(
        select(Device).where(
            Device.device_uid == body.device_uid,
            Device.owner_user_id == user.id,
        )
    )
    device = res.scalar_one_or_none()
    if device is None:
        raise HTTPException(status_code=404, detail="device not found")

    from app.api.v1.telemetry import _bridge_telemetry_to_variables

    await _bridge_telemetry_to_variables(
        device_id=device.id,
        device_uid=device.device_uid,
        event_type="simulator",
        payload={body.variable_key: body.value},
    )

    # Update device last_seen
    device.last_seen_at = datetime.now(timezone.utc)
    await db.commit()

    return {"ok": True, "device_uid": body.device_uid, "variable_key": body.variable_key, "value": body.value}
